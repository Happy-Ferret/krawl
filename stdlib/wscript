def build(bld):
	libs = 'helloworld string'.split()
	for l in libs:
		src = bld.path.ant_glob('std/%s/*.krl' % l)
		bld(
			features     = 'krawl',
			hash_uid     = "krawl standard library: %s" % l,
			source       = src,
			target       = 'std/%s' % l,
			target_brl   = 'std/%s.brl' % l,
			includes     = '.',
			package      = l,
			install_path = '${PREFIX}/include/krawl/std'
		)

	lib = ['std/%s' % l for l in libs]
	bld(
		features     = 'c cstlib',
		target       = 'stdkrawl',
		use          = lib,
		install_path = '${PREFIX}/lib/krawl'
	)
