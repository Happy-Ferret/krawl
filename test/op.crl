import "stdio.h";
import "stdlib.h";

func assert(cond bool)
{
	if !cond {
		stdio.printf("fail\n");
		stdlib.exit(1);
	}
	stdio.printf("ok\n");
}

func main(argc int, argv **byte) int
{
	{
		// random math ops
		a := 5;
		b := 8;
		c := (((((a + b) * 4) / 2 - 6) << 7) % 113) >> 2;
		assert(c == 18);
	}
	{
		// unsigned overflow
		a := 245.(uint8);
		b := 15.(uint8);
		c := a + b;
		assert(c == 4);
	}
	{
		// signed overflow
		a := -125.(int8);
		b := 20.(int8);
		c := a - b;
		assert(c == 111);
	}
	{
		pi := 3.1415;
		halfpi := pi / 2;
		// TODO: varargs type promotion
		stdio.printf("%f / 2 = %f\n", pi, halfpi);
	}
	{
		// some logic operations
		a := 10 > 5 && 2 != 3 && (77-7) == 70;
		b := 10 < 5 || 7 > 0;
		c := a && !b;
		assert(!c);
	}
	{
		ten, five := 10, 5;
		A := five <= ten || ten == five;
		B := five == 5 || ten == 5;
		C := A && B;
		assert(C);
	}
	{
		a := 50;
		b := 50.(uint8);
		c := 50.(uint16);
		a, b, c = ^a, ^b, ^c;
		assert(a == -51 && b == 205 && c == 65485);
	}
	return 0;
}
