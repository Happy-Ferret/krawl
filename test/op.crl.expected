ok
ok
ok
3.141500 / 2 = 1.570750
ok
ok
ok
